research/minigo/dual_net.py

```diff
@@ -0,0 +1,209 @@
+# Copyright 2018 MLBenchmark Group. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""Contains utility and supporting functions for DualNet.
+
+This module provides the model interface, including functions for DualNet model
+bootstrap, training, validation, loading and exporting.
+"""
+
+# pylint: disable=g-bad-import-order
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+
+import tensorflow as tf
+
+import dualnet_model
+import features
+import go
+import model_utils
+import preprocessing
+import symmetries
+
+
+class DualNetwork(object):
+  """The DualNetwork class for the complete model with graph and weights.
+
+  This class can restore the model from saved files, and provide inference for
+  given examples.
+  """
```

Comment:
Why does this exist? It seems to be used for the match, but why is estimator insufficient there? Please comment as to how this differs from the code below for bootstrap, eval, etc.

Reply:
To restore a saved model, estimator needs to restore both the graph definition and the variables, right? I thin here DualNet tries to isolate graph initialization and weights initialization as two models in a play match share the same graph model but with different weights, thus is more efficient.